index/segment/mem/segment_simple.go

```diff
@@ -0,0 +1,171 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+
+	"github.com/uber-go/atomic"
+)
+
+var (
+	errUnknownDocID = errors.New("unknown DocID specified")
+)
+
+// TODO(prateek): investigate impact of native heap
+type simpleSegment struct {
+	opts     Options
+	id       segment.ID
+	docIDGen *atomic.Uint32
+
+	// internal docID -> document
+	docs struct {
+		sync.RWMutex
+		values []document
+	}
+
+	// field (Name+Value) -> postingsManagerOffset
+	termsDict termsDictionary
+
+	searcher searcher
+	// TODO(prateek): add a delete documents bitmap to optimise fetch
+}
+
+type document struct {
+	doc.Document
+	docID      segment.DocID
+	tombstoned bool
+}
+
+// New returns a new in-memory index.
+func New(id segment.ID, opts Options) (Segment, error) {
+	seg := &simpleSegment{
+		opts:      opts,
+		id:        id,
+		docIDGen:  atomic.NewUint32(0),
+		termsDict: newSimpleTermsDictionary(opts),
+	}
+	seg.docs.values = make([]document, opts.InitialCapacity())
+	searcher := newSequentialSearcher(seg, differenceNegationFn, opts.PostingsListPool())
+	seg.searcher = searcher
+	return seg, nil
+}
+
+// TODO(prateek): consider copy semantics for input data, esp if we can store it on the native heap
+func (i *simpleSegment) Insert(d doc.Document) error {
+	return i.insertDocument(document{
+		Document: d,
+		docID:    segment.DocID(i.docIDGen.Inc()),
+	})
+}
+
+func (i *simpleSegment) ensureSufficientCapacity(n int64) {
+	// can early terminate if we have sufficient capacity
+	i.docs.RLock()
+	size := len(i.docs.values)
+	i.docs.RUnlock()
+	if int64(size) > n {
+		return
+	}
+
+	// need to expand capacity
+	i.docs.Lock()
+	size = len(i.docs.values)
+	// expanded since we released the lock
+	if int64(size) > n {
+		i.docs.Unlock()
+		return
+	}
+
+	docs := make([]document, 2*(size+1))
+	copy(docs, i.docs.values)
+	i.docs.values = docs
+	i.docs.Unlock()
+}
+
+func (i *simpleSegment) insertDocument(doc document) error {
+	i.ensureSufficientCapacity(int64(doc.docID))
```

Comment:
Combine the RLocks together perhaps so just need to acquire once?

Reply:
done